### PR TITLE
Binomial generic support

### PR DESCRIPTION
**Pull request by xukai92**
Support

- `binompdf(n::Real, p::Real, k::Number)`, and
- `binomlogpdf(n::Real, p::Real, k::Number)`

as call-backs functions when `k` is neither `Float64` nor `Int`

### CODE REVIEW

File: src/distrs/binom.jl [hunk 1]
```diff
@@ -11,3 +11,21 @@ import .RFunctions:
     binominvccdf,
     binominvlogcdf,
     binominvlogccdf
+
+# pdf for numbers with generic types
+binompdf(n::Real, p::Real, k::Number) = begin
+    # NOTE: fail to do below
+    #       prod(map(i -> (n + 1 - i) / i, 1:Int(k.))) * p^k * (1 - p)^(n - k)
+    n_choose_k = 1; i = 1;
+    while i <= k n_choose_k *= (n + 1 - i) / i; i += 1 end
```
Comment (andreasnoack) [inline src/distrs/binom.jl]: This can overflow for large `n` and `k`. Probably better to compute this as `exp(binomlogpdf)`

File: src/distrs/binom.jl [hunk 2]
```diff
@@ -11,3 +11,21 @@ import .RFunctions:
...
+    n_choose_k * p^k * (1 - p)^(n - k)
+end
+
+# logpdf for numbers with generic types
+binomlogpdf(n::Real, p::Real, k::Number) = begin
```
Comment (andreasnoack) [inline src/distrs/binom.jl]: Why is `k` only restricted to `Number` here?

File: src/distrs/binom.jl [hunk 3]
```diff
@@ -11,3 +11,21 @@ import .RFunctions:
...
+    # NOTE: fail to do below
+    #       sum(map(i -> log((n + 1 - i) / i), 1:k)) + k * log(p) + (n - k) * log(1 - p)
+    log_n_choose_k = 0; i = 1;
+    while i <= k log_n_choose_k += log((n + 1 - i) / i); i += 1 end
+    log_n_choose_k + k * log(p) + (n - k) * log(1 - p)
```
Comment (andreasnoack) [inline src/distrs/binom.jl]: `log(1-p)` -> `log1p(p)` but the best solution would probably be to make `beta` work for `Dual` and [here](https://github.com/JuliaDiff/DiffBase.jl/blob/master/src/rules.jl) with the derivative rules https://www.wolframalpha.com/input/?i=d%2Fda+beta(a,b)

File: src/distrs/binom.jl (hunk 1 above)
Comment (xukai92) [inline src/distrs/binom.jl]: I see!

File: src/distrs/binom.jl (hunk 2 above)
Comment (xukai92) [inline src/distrs/binom.jl]: I just followed the convention in https://github.com/JuliaStats/StatsFuns.jl/blob/master/src/distrs/norm.jl.

File: src/distrs/binom.jl (hunk 3 above)
Comment (xukai92) [inline src/distrs/binom.jl]: I see. For the 2nd point, do you mean it's better to implement d/da beta and d/db beta s.t. `Dual` can use them directly instead of pass things down?

Comment (andreasnoack) [inline src/distrs/binom.jl]: If beta works with dual then you can use the beta function here in the binomial pdf

Comment (andreasnoack) [inline src/distrs/binom.jl]: I decided to add rules for `beta` and `lbeta` to DiffBase to figure out how it works so you can use `lbeta` here instead of rolling your own binomial coefficient.

File: src/distrs/binom.jl (hunk 2 above)
Comment (andreasnoack) [inline src/distrs/binom.jl]: I see. I guess it is left over from the time when `Dual`s weren't `Real`. Should probably be `Real` here.

File: src/distrs/binom.jl (hunk 3 above)
Comment (xukai92) [inline src/distrs/binom.jl]: OK cool. Then do I need to wait for the PR to be accepted?

File: src/distrs/binom.jl (hunk 2 above)
Comment (xukai92) [inline src/distrs/binom.jl]: Do you want me to change all of them (there are other in those call-backs I implemented before)?

File: src/distrs/binom.jl (hunk 3 above)
Comment (andreasnoack) [inline src/distrs/binom.jl]: You can just start adjusting the PR. I'm confident it will get merged very soon.

File: src/distrs/binom.jl (hunk 2 above)
Comment (andreasnoack) [inline src/distrs/binom.jl]: It would be great if you could do that as part of a separate PR.

**Comment by xukai92** [comment]
I've made those changes required.

File: src/distrs/binom.jl [hunk 4]
```diff
@@ -13,19 +13,9 @@ import .RFunctions:
...
 
 # logpdf for numbers with generic types
-binomlogpdf(n::Real, p::Real, k::Number) = begin
-    # NOTE: fail to do below
-    #       sum(map(i -> log((n + 1 - i) / i), 1:k)) + k * log(p) + (n - k) * log(1 - p)
-    log_n_choose_k = 0; i = 1;
-    while i <= k log_n_choose_k += log((n + 1 - i) / i); i += 1 end
-    log_n_choose_k + k * log(p) + (n - k) * log(1 - p)
+binomlogpdf(n::Real, p::Real, k::Real) = begin
+    -log(n + 1) - lbeta(n - k + 1, k + 1) + k * log(p) + (n - k) * log(1 - p)
```
Comment (andreasnoack) [inline src/distrs/binom.jl]: `log(n+1)` -> `log1p(n)` (less important) and `log(1-p)` -> `log1p(-p)` (quite important)

File: src/distrs/binom.jl [hunk 5]
```diff
@@ -13,19 +13,9 @@ import .RFunctions:
...
+binompdf(n::Real, p::Real, k::Real) = exp(binomlogpdf(n, p, k))
 
 # logpdf for numbers with generic types
-binomlogpdf(n::Real, p::Real, k::Number) = begin
-    # NOTE: fail to do below
-    #       sum(map(i -> log((n + 1 - i) / i), 1:k)) + k * log(p) + (n - k) * log(1 - p)
-    log_n_choose_k = 0; i = 1;
-    while i <= k log_n_choose_k += log((n + 1 - i) / i); i += 1 end
-    log_n_choose_k + k * log(p) + (n - k) * log(1 - p)
+binomlogpdf(n::Real, p::Real, k::Real) = begin
```
Comment (andreasnoack) [inline src/distrs/binom.jl]: maybe just remove the `begin` and `end`

File: src/distrs/binom.jl (hunk 4 above)
Comment (xukai92) [inline src/distrs/binom.jl]: Yes!

**Comment by xukai92** [comment]
Thx for helping through :smile:

**Comment by andreasnoack** [comment]
No problem. Thanks for improving the package.

**Comment by andreasnoack** [comment]
Seems like we need a release of ForwardDiff

**Comment by andreasnoack** [comment]
Sorry for the delay here. It seems that we also need a new version of Compat before StatsFuns work on master. https://github.com/JuliaLang/Compat.jl/pull/394

**Comment by nalimilan** [comment]
A new Compat version has been merged now, should be OK to rebase and merge?
